packages/api-plugin-promotions/src/watchers/setPromotionState.js

```diff
@@ -46,30 +54,50 @@ async function markActive(context) {
  * @return {Promise<number>} - The total number of records updated
  */
 async function markCompleted(context) {
-  const { collections: { Promotions } } = context;
+  const { appEvents, collections: { Promotions } } = context;
   const shopTimes = await getCurrentShopTime(context);
   let totalUpdated = 0;
+  const updatePromises = [];
   for (const shop of Object.keys(shopTimes)) {
     const shopTime = shopTimes[shop];
     // eslint-disable-next-line no-await-in-loop
-    const { modifiedCount } = await Promotions.updateMany({
+    const toMarkCompleted = await Promotions.find({
       shopId: shop,
       state: "active",
       endDate: { $lt: shopTime }
-    }, { $set: { state: "completed" } });
-    totalUpdated += modifiedCount;
+    }).toArray();
+    for (const promotion of toMarkCompleted) {
+      appEvents.emit("promotionCompleted", promotion._id);
+      totalUpdated += 1;
+      const updatePromise = Promotions.updateOne({ _id: promotion._id }, { $set: { state: "completed" } });
+      updatePromises.push(updatePromise);
+    }
   }
+  await Promise.all(updatePromises);
   return totalUpdated;
 }
 
 /**
- * @summary capture and change all promotion records who's state should have changed
+ * @summary return closure of markPromotion states with context enclosed
  * @param {Object} context - The application context
- * @return {Promise<Object>} - quantities marked active and completed
+ * @return {Function} - quantities marked active and completed
  */
-export default async function setPromotionState(context) {
-  const totalMadeActive = await markActive(context);
-  const totalMarkedCompleted = await markCompleted(context);
-  Logger.info({ ...logCtx, totalMarkedCompleted, totalMadeActive }, "Scanned promotions for changing state");
-  return { totalMarkedCompleted, totalMadeActive };
+export default function setPromotionState(context) {
+  /**
+   * @summary scan all promotions for any that need to change state
+   * @return {Promise<Object|Number>} - Either an object of completed record counts, or error
+   */
+  async function markPromotionStates() {
+    return new Promise(async (resolve, reject) => {
+      try {
+        const totalMadeActive = await markActive(context);
+        const totalMarkedCompleted = await markCompleted(context);
+        Logger.info({ ...logCtx, totalMarkedCompleted, totalMadeActive }, "Scanned promotions for changing state");
+        resolve({ totalMarkedCompleted, totalMadeActive });
+      } catch (error) {
+        reject(error);
+      }
+    });
```

Comment:
It doesn't look like the `new Promise` wrapper is necessary here. This would be identical:

```js
async function markPromotionStates() {
  const totalMadeActive = await markActive(context);
  const totalMarkedCompleted = await markCompleted(context);
  Logger.info({ ...logCtx, totalMarkedCompleted, totalMadeActive }, "Scanned promotions for changing state");
  return { totalMarkedCompleted, totalMadeActive };
}
```